machine/machine_core/ssh_connection.py

```diff
@@ -205,13 +209,14 @@ def _start_ssh_master(self):
             else:
                 raise exceptions.Failure("SSH master process exited with code: {0}".format(proc.returncode))
 
+        assert proc, "SSH master process not created succesfully"
```

Comment:
This assert is bad and a rather direct result of the retry loop above being a bit of a disaster.  How about lifting that into a separate function which returns the subprocess unconditionally?